Add a way to differentiate restart from stop and keep dhcp config on restart

keszybz opened the pull request:
This is missing two important bits:
- documentation for the new option. This will be mostly a copy&paste job from the existing description of similar options. I'll add that if the general direction in this PR is approved.
- the same mechanism for units which use ExecStop=. I think that we shouldn't add a new option, but instead simply define an environment variable for the ExecStop command, so it can adjust behaviour if wanted. E.g. `SYSTEMD_JOB=stop|restart`. But this might need more design, so I'm ignoring that for now.

dnicolodi commented on units/systemd-networkd.service.in:
```diff
@@ -42,6 +42,7 @@ SystemCallArchitectures=native
 SystemCallErrorNumber=EPERM
 SystemCallFilter=@system-service
 Type=notify
+RestartKillSignal=SIGUSR2
```
Probably just bikeshedding, but I think SIGHUP would be a more conventional choice for this.

ssahani replied:
Yes I think SIGHUP is pretty generic for this . We should do that too.

keszybz replied:
Not really. SIGHUP is conventionally used for "configuration reload". This is about optimizing a full stop-and-start operation to skip some steps if we know that the stop will be followed by a start. In fact, if we ever add support for reloading configuration to networkd without restarting, we'd almost certainly use SIGHUP for this. We shouldn't mix those two operations.

keszybz commented:
I see the test fail badly. Will fix.

yuwata commented:
One of the issue is that `client_send_release()` fails when the client does not have any lease.

keszybz commented:
The crash should be fixed. Let's see how this goes.

yuwata commented:
LGTM.

keszybz commented:
When writing the docs, I realized there's an error in the configuration model: we need to fall back to KillSignal= if RestartKillSignal= is not specified. I changed the code to use an accessor function instead of reading the field directly (and it is unset by default).

I also fixed a bug in the functional tests.

keszybz commented:
@yuwata acked a previous version, and CI passes. Let's merge.
